webgui-new/README.md

```diff
@@ -0,0 +1,76 @@
+# New Next.js (React) based webgui for CodeCompass
+
+## Prerequisites
+
+Make sure you are using at least version 18 of [Node.js](https://nodejs.org/en/).\
+You can also download [nvm](https://github.com/nvm-sh/nvm), and run `nvm install` at the root of the project to use the specified Node version (18.13.0).
+
+## Development
+
+For easier development, first run the webserver (CodeCompass_webserver), then run the development server.\
+Since the webserver and development server are different origins, you have to use the [CORS extension](https://chrome.google.com/webstore/detail/allow-cors-access-control/lhobafahddgcelffkeicbaginigeejlf?hl=en) to resolve CORS issues.
+
+Alternatively, you could add this change to the `MainRequestHandler::begin_request_handler(struct mg_connection* conn_)` method in `mainrequesthandler.cpp`, then build the webserver again.\
+(Note: Do not commit this change, as it is unnecessary in production.)
+
+```c
+mg_send_header(conn_, "Access-Control-Allow-Origin", "http://localhost:3000");
+```
+
+Learn more about CORS [here](https://www.moesif.com/blog/technical/cors/Authoritative-Guide-to-CORS-Cross-Origin-Resource-Sharing-for-REST-APIs/).
+
+Run the webserver with:
+
+```bash
+cd <path-to-webserver-binary> && ./CodeCompass_webserver -w <path-to-workspace-folder> -p <port>
+```
+
+Install npm packages with:
+
+```bash
+npm install
+```
+
+Generate TypeScript files from Thrift by running the following command:
+
+```bash
+chmod +x thrift-codegen.sh && ./thrift-codegen.sh --thrift-source ../
+```
+
+You can specify the properties of the webserver by creating a `.env` file at the root of the project. This is only required for development.\
```

Comment:
Usually the `.env` file is under version control, defining a default configuration. Users can set up a `.env.local` file to override it with local configuration, and that file if ignored from version control. We should also follow this directive.
https://create-react-app.dev/docs/adding-custom-environment-variables/

Reply:
Now with the suggested changes, there is only one `.env` variable, `BACKEND_URL`, which already has a default, fallback value set in `next.config.js`, which is `http://localhost:8080` (because the default port for the webserver is `8080`).

Users can override this by creating a `.env` file and setting the `BACKEND_URL` variable. I understand why using a `.env.local` file may be a good idea, but I don't think it is really necessary in this case, as there is only one variable, which has a fallback value and only required for development.

Of course, if we really want to follow this directive, I can look into this, but in my opinion it is not that necessary in this case.

Reply:
Thanks for refactoring this! 👍 

The goal of the `.env` file is to define the default values, which can be overridden. Of course if you hard-code the default values into `next.config.js`, it becomes an empty file; but the conventional solution would be to place the default value (`http://localhost:8080`) to the `.env` file.

It is a minor issue now and you can leave like this if you want, but most likely it will be refactored later anyway when more environment variables are introduced for future features. Then it would be much better to have all those default values in a single, separate file.

Reply:
@mdeme01 Additional remark: the official Next.js documentation also recommends *NOT* to use runtime configuration in `next.config.js`, like you do, but instead purely environment variables.

https://nextjs.org/docs/api-reference/next.config.js/runtime-configuration
> Note: This feature is considered legacy and does not work with [Automatic Static Optimization](https://nextjs.org/docs/advanced-features/automatic-static-optimization), [Output File Tracing](https://nextjs.org/docs/advanced-features/output-file-tracing#automatically-copying-traced-files), or [React Server Components](https://nextjs.org/docs/advanced-features/react-18/server-components). Please use [environment variables](https://nextjs.org/docs/basic-features/environment-variables) instead in order to avoid initialization overhead.

Reply:
Okay, I removed the public runtime config and commited the `.env` file. Now the variables can be overriden by creating a `.env.local` file and setting the variables inside it.

Because `.env` variables cannot be accessed in the browser, I used the built-in feature of Next.js to expose them, prefixing them with `NEXT_PUBLIC_`.

To determine which URL to use for the configuration, the app needs to know if it is in a production environment or not - I used the `NODE_ENV` variable previously for this, but because I cannot access it in the browser now, I had to create another variable, `NEXT_PUBLIC_DEV_ENV`, which is `false` by default, so in the `.env.local` file, you also have to set this variable to `true` for development.

Reply:
I have used the `http-proxy-middleware` package with environment files for Node.js based proxying to avoid CORS issues in multiple projects, it works.

Such an example is the [TMS](https://gitlab.com/tms-elte/frontend-react) submission management project developed at the ELTE FI in React. See `.env` files and the `src/setupProxy.js` for reference.

I will get on to review the other parts of the PR. If you are stuck with how to implement this, we shall get back to this sometime later. Let's keep this thread open as a reminder.

Reply:
Thank you, I will take a look, but I think you replied to the wrong thread. 🙂

Reply:
Yes, I replied from mobile, and not to the proper thread then :sweat_smile:

Reply:
The value for `NODE_ENV` should be injected during production build or when running the webserver. (Or at least it works like that with vanilla React without NextJS.)

How come it was accessible previously but not now?
Can you link the related part of the codebase?

Reply:
Previously, it was accessible because it was passed as a runtime variable using `publicRuntimeConfig` in `next.config.js`. As that had to be removed, it could no longer be injected, because the client side of the application cannot access environment variables (it will throw the `cannot resolve module` error if you try to use `dotenv`).

However, Next.js has a solution for this, the variables prefixed with `NEXT_PUBLIC_` will be accessible. But you cannot prefix `NODE_ENV`, as it is a built-in variable.

But this no longer applies, as now the webserver URL is set as the current URL both in production and development, because the requests are proxied to the actual webserver in development.